optuna/storages/cached_storage.py

```diff
@@ -0,0 +1,287 @@
+
+        self._storage.set_study_direction(study_id, direction)
+
+    def set_study_user_attr(self, study_id: int, key: str, value: Any) -> None:
+
+        self._storage.set_study_user_attr(study_id, key, value)
+
+    def set_study_system_attr(self, study_id: int, key: str, value: Any) -> None:
+
+        self._storage.set_study_system_attr(study_id, key, value)
+
+    def get_study_id_from_name(self, study_name: str) -> int:
+
+        return self._storage.get_study_id_from_name(study_name)
+
+    def get_study_id_from_trial_id(self, trial_id: int) -> int:
+
+        return self._storage.get_study_id_from_trial_id(trial_id)
+
+    def get_study_name_from_id(self, study_id: int) -> str:
+
+        return self._storage.get_study_name_from_id(study_id)
+
+    def get_study_direction(self, study_id: int) -> StudyDirection:
+
+        return self._storage.get_study_direction(study_id)
+
+    def get_study_user_attrs(self, study_id: int) -> Dict[str, Any]:
+
+        return self._storage.get_study_user_attrs(study_id)
+
+    def get_study_system_attrs(self, study_id: int) -> Dict[str, Any]:
+
+        return self._storage.get_study_system_attrs(study_id)
+
+    def get_all_study_summaries(self) -> List[StudySummary]:
+
+        return self._storage.get_all_study_summaries()
+
+    def create_new_trial(self, study_id: int, template_trial: Optional[FrozenTrial] = None) -> int:
+
+        trial_id, frozen_trial = self._storage.create_new_trial_with_trial(
+            study_id, template_trial
+        )
+        with self._lock:
+            self._study_id_to_trial_id[study_id].append(trial_id)
+            self._trial_id_to_study_id[trial_id] = study_id
+            self._hot_cache[trial_id] = frozen_trial, _TrialUpdates()
+            return trial_id
+
+    def set_trial_state(self, trial_id: int, state: TrialState) -> bool:
+
+        with self._lock:
+            if trial_id in self._hot_cache:
+                cached_trial, updates = self._hot_cache[trial_id]
+                cached_trial.state = state
+                updates.state = state
+                if trial_id in self._dirty_trials:
+                    self._dirty_trials.remove(trial_id)
+                ret = self._update_storage(trial_id)
+                if cached_trial.state.is_finished():
+                    del self._hot_cache[trial_id]
+                return ret
+
+        return self._storage.set_trial_state(trial_id, state)
+
+    def set_trial_param(
+        self,
+        trial_id: int,
+        param_name: str,
+        param_value_internal: float,
+        distribution: distributions.BaseDistribution,
+    ) -> bool:
+
+        with self._lock:
+            if trial_id in self._hot_cache:
+                cached_trial, updates = self._hot_cache[trial_id]
+                study_id = self._trial_id_to_study_id[trial_id]
+                cached_dist = self._param_distributions[study_id].get(param_name, None)
```

Comment:
Since `RDBStorage` performs the distribution compatibility check only at trial-level (see below), it seems that `CachedStorage` also can do it the same level and remove the `_param_distributions` field.
https://github.com/optuna/optuna/blob/5c07d0afffeb506866860fa2ca21a71a4199c075/optuna/storages/rdb/storage.py#L547-L549

```suggestion
                cached_dist = cached_trial.distributions.get(param_name, None)
```

Reply:
The model class checks the distribution compatibility in the study level. (And both of current PR and your suggestion are incompatible with the current implementation.)

Reply:
[This line](https://github.com/optuna/optuna/blob/5c07d0afffeb506866860fa2ca21a71a4199c075/optuna/storages/rdb/models.py#L400) checks the study-level distribution compatibility in RDB storage.

Reply:
I fixed this PR so that it checks distribution compatibility across workers.

Reply:
I see. Thank you for the explanation!